components/external_child_process/child_monitor.cc

```diff
@@ -0,0 +1,171 @@
+/* Copyright (c) 2021 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/external_child_process/child_monitor.h"
+
+#if defined(OS_POSIX)
+#include <errno.h>
+#include <fcntl.h>
+#include <signal.h>
+#include <sys/types.h>
+#include <sys/wait.h>
+#endif
+
+#include <utility>
+
+#include "base/bind_post_task.h"
+#include "base/logging.h"
+#include "base/process/kill.h"
+#include "base/single_thread_task_runner.h"
+#include "base/task/post_task.h"
+#include "base/threading/thread_task_runner_handle.h"
+
+namespace brave {
+
+namespace {
+
+#if defined(OS_POSIX)
+int pipehack[2];
+
+static void SIGCHLDHandler(int signo) {
+  int error = errno;
+  char ch = 0;
+  (void)write(pipehack[1], &ch, 1);
+  errno = error;
+}
+
+static void SetupPipeHack() {
+  if (pipe(pipehack) == -1)
+    VLOG(0) << "pipehack errno:" << errno;
+
+  int flags;
+  for (size_t i = 0; i < 2; ++i) {
+    if ((flags = fcntl(pipehack[i], F_GETFL)) == -1)
+      VLOG(0) << "get flags errno:" << errno;
+    // Nonblock write end on SIGCHLD handler which will notify monitor thread
+    // by sending one byte to pipe whose read end is blocked and wait for
+    // SIGCHLD to arrives to avoid busy reading
+    if (i == 1)
+      flags |= O_NONBLOCK;
+    if (fcntl(pipehack[i], F_SETFL, flags) == -1)
+      VLOG(0) << "set flags errno:" << errno;
+    if ((flags = fcntl(pipehack[i], F_GETFD)) == -1)
+      VLOG(0) << "get fd flags errno:" << errno;
+    flags |= FD_CLOEXEC;
+    if (fcntl(pipehack[i], F_SETFD, flags) == -1)
+      VLOG(0) << "set fd flags errno:" << errno;
+  }
+
+  struct sigaction action;
+  memset(&action, 0, sizeof(action));
+  action.sa_handler = SIGCHLDHandler;
+  sigaction(SIGCHLD, &action, NULL);
+}
+
+static void TearDownPipeHack() {
+  struct sigaction action;
+  memset(&action, 0, sizeof(action));
+  action.sa_handler = SIG_DFL;
+  sigaction(SIGCHLD, &action, NULL);
+  for (size_t i = 0; i < 2; ++i)
```

Comment:
since there will only ever be two pipe elements a loop doesn't add much.

Reply:
fixed in fcf4957f689f39e8f09a0c9aa56494935fcbfd7d